pkg/controller/spinnakerservice/status.go

```diff
@@ -25,81 +27,129 @@ const (
+				if cs.State.Terminated != nil {
+					s.evtRecorder.Eventf(instance, v1.EventTypeWarning, "DeployInProgress", "Pod %s is in Phase: %s. Message: %s", p.Name, p.Status.Phase, cs.State.Terminated.Reason)
+					return Updating, nil
 				}
+				if cs.State.Waiting != nil {
+					s.evtRecorder.Eventf(instance, v1.EventTypeWarning, "DeployInProgress", "Pod %s is in Phase: %s. Message: %s", p.Name, p.Status.Phase, cs.State.Waiting.Reason)
+					return Updating, nil
+				}
+				if !cs.Ready {
+					s.evtRecorder.Eventf(instance, v1.EventTypeWarning, "DeployInProgress", "Pod %s is in Phase: %s. Message: %s", p.Name, p.Status.Phase, p.Status.Reason)
+					return Updating, nil
+				}
+			}
+			break
+		case v1.PodPending:
+			timeOut, err := s.k8sLookup.HasExceededMaxWaitingTime(instance, p)
+			if err != nil {
+				return Failure, err
 			}
-			if string(ac.Type) == "" {
-				if string(fc.Type) != "" && fc.Status == v1.ConditionTrue {
-					log.Info(fmt.Sprintf("Status: Failure, deployment %s has no available condition but has failure condition: %s", it.ObjectMeta.Name, fc.Message))
-					status.Status = Failure
-				} else {
-					log.Info(fmt.Sprintf("Status: Unavailable, deployment %s still has not reported available condition", it.ObjectMeta.Name))
-					status.Status = Unavailable
+			if timeOut {
+				return Failure, nil
+			}
+			for _, cs := range p.Status.ContainerStatuses {
+				if cs.State.Waiting != nil {
+					if "ContainerCreating" == cs.State.Waiting.Reason {
+						s.evtRecorder.Eventf(instance, v1.EventTypeWarning, "DeployInProgress", "Pod %s is in Phase: %s. Message: %s", p.Name, p.Status.Phase, cs.State.Waiting.Reason)
+						return Updating, nil
+					}
+					s.evtRecorder.Eventf(instance, v1.EventTypeWarning, "DeployFailed", "Pod %s has not been able to reach a healthy state is in Phase: %s. Message: %s", p.Name, p.Status.Phase, cs.State.Waiting.Reason)
+					return Failure, nil
 				}
-			} else if ac.Status != v1.ConditionTrue {
-				log.Info(fmt.Sprintf("Deployment %s is available: %s. Message: %s", it.ObjectMeta.Name, ac.Status, ac.Message))
-				if string(fc.Type) != "" && fc.Status == v1.ConditionTrue {
-					status.Status = Failure
-				} else {
-					status.Status = Updating
+				if !cs.Ready {
+					s.evtRecorder.Eventf(instance, v1.EventTypeWarning, "DeployInProgress", "Pod %s is in Phase: %s. Message: %s", p.Name, p.Status.Phase, p.Status.Reason)
+					return Updating, nil
 				}
 			}
-			svcs = append(svcs, st)
-		}
-		status.Services = svcs
-	}
-	status.ServiceCount = len(list.Items)
-	// Go through the list
-	return s.client.Status().Update(context.Background(), svc)
-}
-
-// getSpinnakerServiceImageFromDeployment returns the name of the image
-func (s *statusChecker) getSpinnakerServiceImageFromDeployment(p v1.PodSpec) string {
-	for _, c := range p.Containers {
-		if strings.HasPrefix(c.Name, "spin-") {
-			return c.Image
+			break
+		case v1.PodFailed, v1.PodUnknown:
```

Comment:
This is neat, I didn't know you could collapse case statements like this!

Reply:
😃